rust/src/daemon.rs

```diff
@@ -131,6 +149,139 @@ fn deployment_populate_variant_origin(
+            msg: format!("Unexpected message: {o:?}"),
+        },
+    };
+    write_message(&client, reply).context("Writing client reply")?;
+    tracing::debug!("Acknowleged client");
+    Ok(())
+}
+
+static SHUTDOWN_SIGNAL: Lazy<Mutex<Option<Sender<()>>>> = Lazy::new(|| Mutex::new(None));
+
+fn run_acknowledgement_worker(listener: OwnedSocketlike) {
+    tracing::debug!("Processing clients...");
+    loop {
+        let sock = match rustix::net::accept(&listener) {
+            Ok(s) => s,
+            Err(e) => {
+                tracing::warn!("Failed to accept client: {e}");
+                continue;
+            }
+        };
+        std::thread::spawn(move || {
+            if let Err(e) = client_hello(sock.into_socketlike(), Ok(())) {
+                tracing::warn!("error acknowledging client: {e}");
+            }
+        });
+    }
+}
+
+/// Ensure all asynchronous tasks in this Rust half of the daemon code are stopped.
+/// Called from C++.
+pub(crate) fn daemon_terminate() {
+    if let Some(chan) = (*SHUTDOWN_SIGNAL).lock().unwrap().take() {
+        let _ = chan.send(());
+    }
+}
+
+fn process_one_client(listener: OwnedSocketlike, e: anyhow::Error) -> Result<()> {
+    let incoming = rustix::net::accept(&listener)?;
+    // Send the error to the client
+    client_hello(incoming.into_socketlike(), Err(e))?;
+    // We've successfully sent the error; the caller of this function will
+    // propagate the error.
+    Ok(())
+}
+
+/// Perform initialization steps required by systemd service activation.
+///
+/// This ensures that the system is running under systemd, then receives the
+/// socket-FD for main IPC logic, and notifies systemd about ready-state.
+pub(crate) fn daemon_main(debug: bool) -> Result<()> {
+    let handle = tokio::runtime::Handle::current();
+    let _tokio_guard = handle.enter();
+    if !systemd::daemon::booted()? {
+        return Err(anyhow!("not running as a systemd service"));
+    }
+    let init_res: Result<()> = crate::ffi::daemon_init_inner(debug).map_err(|e| e.into());
+    tracing::debug!("Initialization result: {init_res:?}");
+
+    let mut fds = systemd::daemon::listen_fds(false)?.iter();
+    let fd = fds.next().ok_or_else(|| anyhow!("Missing socket"))?;
+    if fds.next().is_some() {
+        return Err(anyhow!("Expected exactly 1 fd from systemd activation"));
+    }
+    tracing::debug!("Initializing from socket activation; fd={fd}");
+    let listener = unsafe { OwnedFd::from_raw_fd(fd) }.into_socketlike();
+    // In the socket case, we will process the initialization error later.
+
+    // On success, we spawn a helper task that just responds with
+    // sucess to clients that connect via the socket.  In the future,
+    // perhaps we'll expose an API here.
+    tracing::debug!("Spawning acknowledgement task");
+    match init_res {
+        Ok(()) => {
+            dbg!(&listener);
+            std::thread::spawn(move || run_acknowledgement_worker(listener));
```

Comment:
Also now that we're using a streaming protocol we can switch all of this to use a proper async tokio task instead of a thread.  But, not a blocker.

Reply:
so instead of spawning thread to run task concurrenty we use `tokio::task::spawn(async move {run_acknowledgement_worker(listener).await })`. But wouldn't this require the `daemon_main` to be async or do we want to block on this with `futures::executor::block_on(async {...})`